src/notification.c

```diff
@@ -234,129 +234,184 @@ check_value_type (const char *key,
 }
 
 static gboolean
-check_priority (GVariant *value,
-                GError **error)
+parse_priority (GVariantBuilder  *builder,
+                GVariant         *value,
+                GError          **error)
 {
   const char *priorities[] = { "low", "normal", "high", "urgent", NULL };
+  const char *priority;
 
   if (!check_value_type ("priority", value, G_VARIANT_TYPE_STRING, error))
     return FALSE;
 
-  if (!g_strv_contains (priorities, g_variant_get_string (value, NULL)))
-    {
-      g_set_error (error,
-                   XDG_DESKTOP_PORTAL_ERROR,
-                   XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
-                   "%s not a priority", g_variant_get_string (value, NULL));
-      return FALSE;
-    }
+  priority = g_variant_get_string (value, NULL);
+
+  if (g_strv_contains (priorities, priority))
+    g_variant_builder_add (builder, "{sv}", "priority", value);
+  else
+    g_debug ("Unsupported priority %s filtered from notification", priority);
 
   return TRUE;
 }
 
 static gboolean
-check_button (GVariant *button,
-              GError **error)
+parse_button (GVariantBuilder  *builder,
+              GVariant         *button,
+              GError          **error)
 {
   int i;
-  gboolean has_label = FALSE;
-  gboolean has_action = FALSE;
+  g_autoptr(GVariant) label = NULL;
+  g_autoptr(GVariant) action = NULL;
+  g_autoptr(GVariant) target = NULL;
 
   for (i = 0; i < g_variant_n_children (button); i++)
     {
       const char *key;
       g_autoptr(GVariant) value = NULL;
 
       g_variant_get_child (button, i, "{&sv}", &key, &value);
+
       if (strcmp (key, "label") == 0)
-        has_label = TRUE;
+        {
+          if (!check_value_type (key, value, G_VARIANT_TYPE_STRING, error))
+            return FALSE;
+          label = g_steal_pointer (&value);
```

Comment:
this is leaking when there are multiple "label" entries. some for the other keys. either `g_clear_pointer (&label, g_variant_unref)` or change the if before to check for `label == NULL` depending on if you want to use the first or last occurrence of the key.

Reply:
Seeing the changed code: is it documented what happens with multiple keys? Should this maybe even return an error?

Reply:
The documentation of the portal doesn't mention anything so this seems to be undefined and the code should be fine. Either way, we should update the spec to either throw an error when a key exists multiple times (which might not be possible if there are actual apps which already do this), or at least update if the first or last occurrence of a key is used.

Reply:
not sure, the docs don't say anything about it, and the old code just hands what ever keys are present to the backend. The gtk backend simple passes everything to GNOME Shell or uses `g_variant_lookup_value()` when using FDO. The docs for ` g_variant_lookup_value()`  don't specify a behavior, but a quick look at the code it seams that it just returns the first found value.

Reply:
Alright, this is fine then.

It would still be good to define what happens with multiple keys in a dict for all portals (first or last occurrence wins) but let's not block this MR on it.